### PR TITLE
netif: Initial import of ng_* version

### PR DESCRIPTION
**Pull request by authmillenon**
~~Depends on #2397~~ It does not

### CODE REVIEW

**Comment by haukepetersen** [comment]
I am not quite sure what the motivation behind this module (in the current form) is. Is this intended to act as central registry for network devices, as alternative to including the handling of these also into `netreg`?

**Comment by authmillenon** [comment]
The motivation is to have a central registry of interfaces, that can be returned via the FIB. The reason I did not put it into `netreg` is two-fold:

1. Nostalgia: this module is pretty much the ng version of `net_if`
2. I don't see much connection in the functionality, other than registering PIDs somewhere. `netreg` is more a "I'm interested in packets of a certain type", `netif` is more "I want to be an endpoint for routing. Network layer what do you need of me." (Last sentence reflects the add/remove callbacks).

**Comment by authmillenon** [comment]
See it this way: you can leave `netif` out if you don't want routing in theory, but you can not leave out `netreg`, since it is integral to the inter-layer communication.

**Comment by cgundogan** [comment]
>ERROR: The following new files generate doxygen warnings: sys/include/net/ng_netif.h

**Comment by authmillenon** [comment]
Fixed

**Comment by haukepetersen** [comment]
I think I understand. So this module would also be used when I want to get a list of network devices for implementing shell commands, right?

**Comment by authmillenon** [comment]
Yes.

File: sys/net/crosslayer/ng_netreg/Makefile [hunk 1]
```diff
@@ -0,0 +1 @@
+include $(RIOTBASE)/Makefile.base
```
Comment (haukepetersen) [inline sys/net/crosslayer/ng_netreg/Makefile]: this file is unrelated, right?

File: sys/include/net/ng_netif.h [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+#define NG_NETIF_NUMOF  (1)
+#endif
+
+/**
+ * @brief   The add/remove operation to set network layer protocol
+ *          specific options for an interface.
+ *
+ * @param[in] pid   The PID to the new interface.
+ */
+typedef void (*ng_netif_op_t)(kernel_pid_t pid);
```
Comment (haukepetersen) [inline sys/include/net/ng_netif.h]: can this be removed? Its not used in the implementation...

File: sys/include/net/ng_netif.h [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ * @brief   The add and remove handlers to set network layer protocol
+ *          specific options for an interface.
+ *
+ * @details If you implement a pair, please add it to the list in ng_netif.c
+ *          statically.
+ */
+typedef struct {
+    ng_netif_op_t add;      /**< The add operation */
+    ng_netif_op_t remove;   /**< The remove operation */
+} ng_netif_handler_t;
```
Comment (haukepetersen) [inline sys/include/net/ng_netif.h]: same for this type...

File: sys/include/net/ng_netif.h (hunk 2 above)
Comment (haukepetersen) [inline sys/include/net/ng_netif.h]: sorry, overlooked the usage, never mind my comment above...

File: sys/net/crosslayer/ng_netif/ng_netif.c [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+ * @file
+ */
+
+#include <stdio.h>
+
+#include <errno.h>
+#include "kernel_types.h"
+#include "net/ng_netif.h"
+
+static ng_netif_handler_t if_handler[] = {
```
Comment (haukepetersen) [inline sys/net/crosslayer/ng_netif/ng_netif.c]: I am not sure if we need this handlers. To add/set/rem e.g. an IPv6 address for a device, I would do it the following way (pseudo code):
```
dev_pid = ...;    // PID of the network device
addr = {};        // IPv6 address to add to the device
// get the PID for the IPv6 protocol
ip_pid = netreg_lookup(NETTYPE_IPV6);
// then I set a new address using netapi
netapi_set(ip_pid, NETCONF_OPT_ADDRESS, dev_pid, &addr, sizeof(addr));
```
@authmillenon Do you agree with this?

Comment (authmillenon) [inline sys/net/crosslayer/ng_netif/ng_netif.c]: No. There are options that *must* be set on an interface, when it is used with IPv6 etc. Additionaly, there are actions that are done upon an interface if it is initialized, e.g. sending out neighbor solicitations, auto-address config packets, or DHCP packtes. These handlers bootstrap this behavior nicely if the interface is added post-IPv6/IPv4/… initialization. If they are initialized before-hand, the network protocol can just do it on it's own initialization.

Comment (authmillenon) [inline sys/net/crosslayer/ng_netif/ng_netif.c]: Addresses are the smallest problem… though I'm still not sure how we handle lists of them with `netapi_set`, since an IPV6 interface always has multiple addresses.

Comment (authmillenon) [inline sys/net/crosslayer/ng_netif/ng_netif.c]: (but let's move that discussion to the point, when I actually implemented IPv6 with the current netapi ;-))

Comment (haukepetersen) [inline sys/net/crosslayer/ng_netif/ng_netif.c]: makes sense. So for now I am fine with this.

File: sys/net/crosslayer/ng_netif/ng_netif.c [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+            for (int j = 0; if_handler[j].add != NULL; j++) {
+                if_handler[j].add(pid);
+            }
+
+            return 0;
+        }
+    }
+
+    return ENOMEM;
```
Comment (haukepetersen) [inline sys/net/crosslayer/ng_netif/ng_netif.c]: maybe negative return value for consistency?

**Comment by haukepetersen** [comment]
just this one thing about the return value and the `ng_netreg/Makefile` that has nothing to do with this PR... I will ack once this is fixed.

**Comment by authmillenon** [comment]
Addressed comments

**Comment by haukepetersen** [comment]
looks good to me -> ACK when squashed and Travis is happy.

**Comment by haukepetersen** [comment]
kicked Travis

**Comment by authmillenon** [comment]
Squashed
